src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        # Find all the output placement rules crated by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferInfo'] is not None \
+           and wflow['TransferInfo']['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            for dataCont in wflow['OutputDatasets']:
+                if gran == 'container':
+                    wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(dataCont, account=rucioAcct))
+                elif gran == 'block':
+                    blocks = self.rucio.getBlocksInContainer(dataCont)
+                    for block in blocks:
+                        wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(block, account=rucioAcct))
+        return wflow
+
+    def cleanRucioRules(self, wflow):
+        """
+        Cleans all the Rules present in the field 'RulesToClean' in the MSRuleCleaner
+        workflow representation. And fills the relevant Cleanup Status.
+        :param wflow:   A MSRuleCleaner workflow representation
+        :return:        wflow
+        """
+        # NOTE: The function should be called independently and sequentially from
+        #       The Input and the respective BlockLevel pipelines.
+
+        # NOTE: The current functional pipeline is always the last one in the PlineMarkers list
+        currPline = wflow['PlineMarkers'][-1]
+        delResutls = []
+        if wflow['TransferInfo'] is not None \
+           and wflow['TransferInfo']['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            if self.isRealMode:
+                for rule in wflow['RulesToClean'][currPline]:
+                    msg = "%s: Deleting rule['id']: %s "
+                    msg += "For did: %s"
+                    self.logger.info(msg, currPline, rule['id'], rule['name'])
+                    delResult = self.rucio.deleteRule(rule['id'])
+                    delResutls.append(delResult)
+                    if not delResult:
+                        msg = "%s: Failed to delete rule['id']: %s "
+                        msg += "For did: %s"
+                        self.logger.warning(msg, currPline, rule['id'], rule['name'])
+            else:
+                for rule in wflow['RulesToClean'][currPline]:
+                    delResutls.append(True)
+                    msg = "%s: Is about to delete rule['id']: %s "
+                    msg += "For did: %s"
+                    self.logger.debug(msg, currPline, rule['id'], rule['name'])
+        else:
+            delResutls.append(False)
+            msg = "%s: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. "
+            msg += "Skip deleting any rules for: %s "
+            self.logger.debug(msg, currPline, wflow['RequestName'])
+
+        # Set the cleanup flag:
+        wflow['CleanupStatus'][currPline] = all(delResutls)
```

Comment:
no rules deleted will set it to True. I guess I'd remake this logic. See comment above.

Reply:
Changes to the code here won't protect us from having positive value when a `no rule was deleted` event happens. If you refer to an `all([])` behavior of this aggregation - I just never expect and empty list to land here because of the explanation given to the previous comment. While in case we want to avoid a list full of positive values when no rule was deleted at all then the change must happen in [2].  
Actually during my tests i did not see a problem with workflows having no rules to be deleted... or I suppose I should expect to have a workflow full of rules that have been already deleted and we hit that Exception for all of them ... but in this case I would consider the workflow already clean. I hope I am not missing something, btw. I will be glad to hear what do you think?

[2] 
 https://github.com/dmwm/WMCore/blob/51e3e2f6975b92abbbd9bf7291cf7e8d8eabd2e3/src/python/WMCore/Services/Rucio/Rucio.py#L719-L720

Reply:
I guess `delResutls` will only be `[]` if there were no rules to be deleted. Thus cleanup flag should go to True indeed.
Otherwise, it will have Falses/Trues according to the status of the rule deletion, which the `all` builtin will properly evaluate. So I think we are good here.